### PR TITLE
Support subclasses of chains for langchain flavor

### PR DESCRIPTION
**Pull request by liangz1**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?
This PR extends the mlflow langchain flavor to support saving and loading all subclasses of Chain.

Before:
| **class of model** | **model log<br>& load<br>(langchain<br>& pyfunc)** | **extract llm<br>type in model<br>metadata** | **model<br>predict** | **pyfunc<br>spark_udf** |
|--------------|--------------|--------------|--------------|--------------|
| LLMChain  | ✅  | ✅  | ✅ | ✅ |
| AgentExecutor  | ✅  | ✅  | ✅ | ✅ |
| other  | ❌  | ❌  |❌ | ❌ |

After:

| **class of model** | **model log<br>& load<br>(langchain<br>& pyfunc)** | **extract llm<br>info in model<br>metadata** | **model<br>predict** | **pyfunc<br>spark_udf** |
|--------------|--------------|--------------|--------------|--------------|
| subclasses of<br>LLMChain  | ✅  | ✅  | ✅ | ✅ |
| subclasses of<br>AgentExecutor  | ✅  | ✅  | ✅ | ✅ |
| subclasses<br>of Chain<br>(see known<br>exceptions<br>below)  | ✅  | ❌ (llm info<br>exists in<br>model.yaml) | ✅* | ✅* |
| Chains<br>containing<br>memory  | ❌  | ❌  |❌ | ❌ |

\* Given all the Chains have some bugs in SerDe, we cannot test any concrete Chains. It should work and we can add tests after langchain fixed the bugs.

Here is a list of Chains that langchain supports loading. 
https://github.com/hwchase17/langchain/blob/0c3de0a0b32fadb8caf3e6d803287229409f9da9/langchain/chains/loading.py#L409
```python
type_to_loader_dict = {
    "api_chain": _load_api_chain, # ❌ SISO requires "requests_wrapper" in kwargs
    "hyde_chain": _load_hyde_chain, # ❌ SISO requires "embeddings" in kwargs
    "llm_chain": _load_llm_chain, # SISO ✅
    "llm_bash_chain": _load_llm_bash_chain, # SISO ❌ Bug, fixing by langchain
    "llm_checker_chain": _load_llm_checker_chain, # ❌ SISO, but contains SequentialChain, which does not support saving
    "llm_math_chain": _load_llm_math_chain, # SISO ❌ Bug, fixing by langchain
    "llm_requests_chain": _load_llm_requests_chain, # ❌ SISO requires "requests_wrapper" in kwargs
    "pal_chain": _load_pal_chain, # SIMO ❌ Bug, reported to langchain ["result", "intermediate_steps"]
    "qa_with_sources_chain": _load_qa_with_sources_chain, # ❌ MIMO [self.input_docs_key, self.question_key] -> [self.answer_key, self.sources_answer_key, "source_documents"]
    "stuff_documents_chain": _load_stuff_documents_chain, # MISO ❌ Bug, fixing by langchain
    "map_reduce_documents_chain": _load_map_reduce_documents_chain, # ❌ MIMO ["result", "intermediate_steps"]
    "map_rerank_documents_chain": _load_map_rerank_documents_chain, # ❌ MIMO ["result", "intermediate_steps"]
    "refine_documents_chain": _load_refine_documents_chain, # ❌ MIMO ["result", "intermediate_steps"]
    "sql_database_chain": _load_sql_database_chain, # ❌ SIMO ["result", "intermediate_steps"], requires "database" in kwargs 
    "vector_db_qa_with_sources_chain": _load_vector_db_qa_with_sources_chain, # ❌ SIMO requires "vectorstore" in kwargs
    "vector_db_qa": _load_vector_db_qa, # ❌ SIMO requires "vectorstore" in kwargs
}
```

Among them, MLflow cannot support the following number of chains and reasons:
1. Multiple Outputs: 8 chains are blocked. mlflow plans to have a follow-up PR to support it.
2. Needs `kwargs` during mlflow.pyfunc.load_model: 6 chains are blocked. mlflow plans to have a follow-up PR to support it.
3. langchain bug: 4 chains are blocked (LLMBashChain, LLMMathChain, StuffDocumentsChain, PALChain). Filed https://github.com/hwchase17/langchain/issues/5131, https://github.com/hwchase17/langchain/issues/5160, https://github.com/hwchase17/langchain/issues/5224

### Known exceptions
Examples of Chain classes that are known to be not supported as of `langchain==0.0.176`:
1. `class ConversationChain(LLMChain): Chain to have a conversation and load context from memory.` Reason: (from `langchain`) Saving of memory is not yet supported.

## How is this patch tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

mlflow langchain flavor allows logging and loading all subclasses of Chain, as long as their serialization / deserialization methods are implemented by langchain.

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by WeichenXu123** [comment]
@liangz1 

Could you also manually test this for dolly ? https://github.com/mlflow/mlflow/issues/8460

(We don't need to add dolly tests in CI)

**Review by minkin-koantek — APPROVED** [review]
worth trying

**Comment by WeichenXu123** [comment]
@minkin-koantek could you help test ?

You can install this branch of the PR via:

```
pip install git+https://github.com/liangz1/mlflow.git@support-more-chains
```

**Comment by minkin-koantek** [comment]
I will test it in the next 12 hr thx!

**Comment by liangz1** [comment]
@WeichenXu123 Thanks for linking the issue.
@minkin-koantek If you don't mind sharing the code to reproduce this issue https://github.com/mlflow/mlflow/issues/8460, I can also help take a look. Thanks!

File: mlflow/langchain/__init__.py [hunk 1]
```diff
@@ -65,7 +65,13 @@
 _TOOLS_DATA_FILE_NAME = "tools.pkl"
 _TOOLS_DATA_KEY = "tools_data"
 _MODEL_TYPE_KEY = "model_type"
-
+_UNSUPPORTED_MODEL_ERROR_MESSAGE = "MLflow langchain flavor only supports logging " + \
+"subclasses of langchain.chains.base.Chain and langchain.agents.agent.AgentExecutor " + \
+"instances, found {instance_type}"
```
Comment (harupy) [inline mlflow/langchain/__init__.py:70]:
```suggestion
_UNSUPPORTED_MODEL_ERROR_MESSAGE = (
    "MLflow langchain flavor only supports logging "
    "subclasses of langchain.chains.base.Chain and langchain.agents.agent.AgentExecutor "
    "instances, found {instance_type}"
```

We can remove `+ \`.

File: mlflow/langchain/__init__.py [hunk 2]
```diff
@@ -293,11 +299,10 @@ def log_model(
     import langchain
 
     if not isinstance(
-        lc_model, (langchain.chains.llm.LLMChain, langchain.agents.agent.AgentExecutor)
+        lc_model, (langchain.chains.Chain, langchain.agents.agent.AgentExecutor)
     ):
         raise mlflow.MlflowException.invalid_parameter_value(
-            "MLflow langchain flavor only supports logging langchain.chains.llm.LLMChain and "
-            + f"langchain.agents.agent.AgentExecutor instances, found {type(lc_model)}"
+            _UNSUPPORTED_MODEL_ERROR_MESSAGE.format(instance_type=str(type(lc_model)))
```
Comment (harupy) [inline mlflow/langchain/__init__.py:305]:
```suggestion
            _UNSUPPORTED_MODEL_ERROR_MESSAGE.format(instance_type=type(lc_model).__name__)
```

Do we need `__name__` here?

```python
>>> str(type(0))
"<class 'int'>"
>>> str(type(0).__name__)
'int'
```

File: tests/langchain/test_langchain_model_export.py [hunk 3]
```diff
@@ -84,14 +95,18 @@ def create_openai_llmagent():
     return initialize_agent(tools, llm, agent=AgentType.ZERO_SHOT_REACT_DESCRIPTION, verbose=True)
 
 
-def create_model(llm_type, model_path=None):
-    if llm_type == "openai":
+def create_model(model_type, model_path=None):
```
Comment (harupy) [inline tests/langchain/test_langchain_model_export.py:98]: Do we need this function? Can we just directly use each function?

Comment (liangz1) [reply tests/langchain/test_langchain_model_export.py:98]: I implemented this function for better readability. It's a centralized place for us to easily know which models are being tested, and hide the implementation details for creating each model.
Do you think it is useful in this sense?

Comment (harupy) [reply tests/langchain/test_langchain_model_export.py:98]: Makes sense, not a big deal. I'd probably remove it for quicker code navigation:

```python
model = create_model("qaevalchain")
# need to jump to create_model first,
# then find a matching if branch
# then jump to the matched function
```

```python
model = create_openai_llmagent()
# can directly jump to `create_openai_llmagent`
```

Comment (liangz1) [reply tests/langchain/test_langchain_model_export.py:99]: Yeah, actually now I also feel it's more convenient to remove this function :)

File: mlflow/langchain/__init__.py [hunk 4]
```diff
@@ -343,9 +351,9 @@ def _save_model(model, path):
     model_data_path = os.path.join(path, _MODEL_DATA_FILE_NAME)
     model_data_kwargs = {_MODEL_DATA_KEY: _MODEL_DATA_FILE_NAME}
 
-    if type(model) == langchain.chains.llm.LLMChain:
+    if isinstance(model, langchain.chains.llm.LLMChain):
         model.save(model_data_path)
-    elif type(model) == langchain.agents.agent.AgentExecutor:
+    elif isinstance(model, langchain.agents.agent.AgentExecutor):
```
Comment (liangz1) [inline mlflow/langchain/__init__.py:356]: @sunishsheth2009 Would all subclasses of AgentExecutor work out-of-the-box? Would you prefer I revert this line of change? Thanks!

Comment (liangz1) [reply mlflow/langchain/__init__.py:356]: Besides, I did not see a test for spark_udf with AgentExecutor. Does it mean we don't support this use case?

**Comment by minkin-koantek** [comment]
> @WeichenXu123 Thanks for linking the issue. @minkin-koantek If you don't mind sharing the code to reproduce this issue #8460, I can also help take a look. Thanks!

The setup is based on a Databricks Dolly script from dbdemos.  If I am unsuccessful in getting the model to register I will isolate it so you can try it

**Comment by minkin-koantek** [comment]
image

The type of my model tested is **langchain.chains.combine_documents.stuff.StuffDocumentsChain**

Code used to deploy:
`def publish_model_to_mlflow(llm_model):

  with mlflow.start_run() as run:
      # Save model to MLFlow
      # Note that this only saves the langchain pipeline (we could also add the ChatBot with a custom Model Wrapper class)
      # See https://mlflow.org/docs/latest/models.html#custom-python-models for an example
      # The vector database lives outside of your model
      mlflow.langchain.log_model(llm_model, artifact_path="model")
      model_registered = mlflow.register_model(f"runs:/{run.info.run_id}/model", "gardening-dolly-7b-bot")

  # Move the model in production
  client = mlflow.tracking.MlflowClient()
  print("registering model version "+model_registered.version+" as production model")
  client.transition_model_version_stage("gardening-dolly-7b-bot", model_registered.version, stage = "Production", archive_existing_versions=True)

def load_model_and_answer(similar_docs, question): 
  # Note: this will load the model once more in memory
  # Load the langchain pipeline & run inferences
  chain = mlflow.pyfunc.load_model(model_uri)
  chain.predict({"input_documents": similar_docs, "human_input": question})`
  
resulting message after fixing promotion code:
`2023/05/19 14:13:33 WARNING mlflow: MLflow does not guarantee support for Chains outside of the subclasses of LLMChain, found StuffDocumentsChain
Registered model 'gardening-dolly-7b-bot' already exists. Creating a new version of this model...
2023/05/19 14:13:37 INFO mlflow.tracking._model_registry.client: Waiting up to 300 seconds for model version to finish creation. Model name: gardening-dolly-7b-bot, version 3
Created version '3' of model 'gardening-dolly-7b-bot'.
registering model version 3 as production model`

File: mlflow/pyfunc/__init__.py [hunk 5]
```diff
@@ -411,8 +411,13 @@ def predict(self, data: PyFuncInput) -> PyFuncOutput:
         :return: Model predictions as one of pandas.DataFrame, pandas.Series, numpy.ndarray or list.
         """
         input_schema = self.metadata.get_input_schema()
+        _logger.debug("Data before _enforce_schema: %s", data)
+        _logger.debug("Input schema: %s", input_schema)
+        # Input schema: ['input_documents': string]
+        # Expected: ['input_documents': string (Document?), 'question': string]
```
Comment (liangz1) [inline mlflow/pyfunc/__init__.py:417]:
```
2023/05/23 11:05:37 DEBUG mlflow.pyfunc: Data before _enforce_schema: [{'input_documents': [Document(page_content='I love MLflow.', metadata={'source': '/path/to/mlflow.txt'}), Document(page_content='I love langchain.', metadata={'source': '/path/to/langchain.txt'}), Document(page_content='I love AI.', metadata={'source': '/path/to/ai.txt'})], 'question': 'What do I like?'}]
2023/05/23 11:05:37 DEBUG mlflow.pyfunc: Input schema: ['input_documents': string]
2023/05/23 11:05:37 DEBUG mlflow.pyfunc: Data after _enforce_schema:                                      input_documents
0  [page_content='I love MLflow.' metadata={'sour...
2023/05/23 11:05:37 DEBUG mlflow: Input pandas data:                                                                                                                                                                                                                 input_documents
0  [page_content='I love MLflow.' metadata={'source': '/path/to/mlflow.txt'}, page_content='I love langchain.' metadata={'source': '/path/to/langchain.txt'}, page_content='I love AI.' metadata={'source': '/path/to/ai.txt'}]
2023/05/23 11:05:37 DEBUG mlflow: messages = [{'input_documents': [Document(page_content='I love MLflow.', metadata={'source': '/path/to/mlflow.txt'}), Document(page_content='I love langchain.', metadata={'source': '/path/to/langchain.txt'}), Document(page_content='I love AI.', metadata={'source': '/path/to/ai.txt'})]}]
2023/05/23 11:05:37 DEBUG mlflow.langchain.api_request_parallel_processor: requests = [{'input_documents': [Document(page_content='I love MLflow.', metadata={'source': '/path/to/mlflow.txt'}), Document(page_content='I love langchain.', metadata={'source': '/path/to/langchain.txt'}), Document(page_content='I love AI.', metadata={'source': '/path/to/ai.txt'})]}]
2023/05/23 11:05:37 DEBUG mlflow: Input pandas data:                                                                                                                                                                                                                 input_documents
0  [page_content='I love MLflow.' metadata={'source': '/path/to/mlflow.txt'}, page_content='I love langchain.' metadata={'source': '/path/to/langchain.txt'}, page_content='I love AI.' metadata={'source': '/path/to/ai.txt'}]
2023/05/23 11:05:37 DEBUG mlflow: messages = [{'input_documents': [Document(page_content='I love MLflow.', metadata={'source': '/path/to/mlflow.txt'}), Document(page_content='I love langchain.', metadata={'source': '/path/to/langchain.txt'}), Document(page_content='I love AI.', metadata={'source': '/path/to/ai.txt'})]}]
2023/05/23 11:05:37 DEBUG mlflow.langchain.api_request_parallel_processor: requests = [{'input_documents': [Document(page_content='I love MLflow.', metadata={'source': '/path/to/mlflow.txt'}), Document(page_content='I love langchain.', metadata={'source': '/path/to/langchain.txt'}), Document(page_content='I love AI.', metadata={'source': '/path/to/ai.txt'})]}]
```

Comment (liangz1) [reply mlflow/pyfunc/__init__.py:417]: It turns out to be a langchain issue. The `chain.input_keys` does not return the correct value.

File: mlflow/langchain/__init__.py (hunk 4 above)
Comment (liangz1) [reply mlflow/langchain/__init__.py:356]: I verified that AgentExecutor works well in spark_udf.

File: tests/langchain/test_langchain_model_export.py [hunk 6]
```diff
@@ -268,11 +268,52 @@ def test_langchain_agent_model_predict():
...
+
+
+def test_langchain_native_log_and_load_qa_with_sources_chain():
+    # StuffDocumentsChain is a subclass of Chain
+    model = create_qa_with_sources_chain()
+    with mlflow.start_run():
+        logged_model = mlflow.langchain.log_model(model, "langchain_model")
+
+    loaded_model = mlflow.langchain.load_model(logged_model.model_uri)
+    assert model == loaded_model
```
Comment (dbczumar) [inline tests/langchain/test_langchain_model_export.py:288]: Can we add at test case for pyfunc predict too?

**Review by dbczumar — APPROVED** [review]
LGTM once `predict()` and `spark_udf()` coverage is added :)

**Comment by dbczumar** [comment]
Thanks @liangz1 !

**Review by harupy — APPROVED** [review]
LGTM!

**Pull request merged**